File: csharp/Teller.cs
using System.Collections.Generic;

namespace supermarket
{

    public class Teller
    {

        private readonly SupermarketCatalog _catalog;
        private readonly List<IOffer> _offers = new List<IOffer>();

        public Teller(SupermarketCatalog catalog)
        {
            this._catalog = catalog;
        }

        public void AddSpecialOffer(SingleProductOffer singleProductOffer)
        {
            this._offers.Add(singleProductOffer);
        }

        public Receipt ChecksOutArticlesFrom(ShoppingCart theCart)
        {
            var receipt = new Receipt();
            var productQuantities = theCart.GetItems();
            foreach (ProductQuantity pq in productQuantities) {
                var p = pq.Product;
                var quantity = pq.Quantity;
                var unitPrice = this._catalog.GetUnitPrice(p);
                var price = quantity * unitPrice;
                receipt.AddProduct(p, quantity, unitPrice, price);
            }
            theCart.HandleOffers(receipt, this._offers, this._catalog);

            return receipt;
        }

    }
}

File: java/src/main/java/dojo/supermarket/model/Teller.java
package dojo.supermarket.model;

import java.util.ArrayList;
import java.util.List;

public class Teller {

    private final SupermarketCatalog catalog;
    private List<Offer> offers = new ArrayList<>();

    public Teller(SupermarketCatalog catalog) {
        this.catalog = catalog;
    }

    public void addSpecialOffer(SpecialOfferType offerType, Product product, double argument) {
        this.offers.add(new SingleProductOffer(offerType, product, argument));
    }

    public void addSpecialOffer(Offer offer) {
        this.offers.add(offer);
    }

    public Receipt checksOutArticlesFrom(ShoppingCart theCart) {
        Receipt receipt = new Receipt();
        List<ProductQuantity> productQuantities = theCart.getItems();
        for (ProductQuantity pq: productQuantities) {
            Product p = pq.getProduct();
            double quantity = pq.getQuantity();
            double unitPrice = this.catalog.getUnitPrice(p);
            double price = quantity * unitPrice;
            receipt.addProduct(p, quantity, unitPrice, price);
        }

        handleOffers(theCart, receipt);

        return receipt;
    }

    private void handleOffers(ShoppingCart theCart, Receipt receipt) {
        for (Offer offer : this.offers) {
            if (offer.canApplyTo(theCart.productQuantities())) {
                receipt.addDiscount(offer.getDiscount(theCart.productQuantities(), catalog));
            }
        }
    }
}
